src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -349,22 +378,36 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
             }
         }
 
-        private int FindEntry(TKey key)
+        private ref Entry FindEntry(TKey key, out bool found)
         {
-            if (key == null)
-            {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
-            }
+            Debug.Assert(key != null);
 
+            found = true;
             if (buckets != null)
             {
-                int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
-                for (int i = buckets[hashCode % buckets.Length]; i >= 0; i = entries[i].next)
+                int hashCode = (comparer == null ? EqualityComparer<TKey>.Default.GetHashCode(key) : comparer.GetHashCode(key)) & 0x7FFFFFFF;
+                int targetBucket = hashCode % buckets.Length;
+                int i = buckets[targetBucket];
+                while (i >= 0)
                 {
-                    if (entries[i].hashCode == hashCode && comparer.Equals(entries[i].key, key)) return i;
+                    ref Entry candidateEntry = ref entries[i];
+                    if (candidateEntry.hashCode == hashCode && ((comparer == null && EqualityComparer<TKey>.Default.Equals(candidateEntry.key, key)) || (comparer != null && comparer.Equals(candidateEntry.key, key))))
```

Comment:
Currently it checks `comparer` twice:
```ASM
       test     rcx, rcx
       jne      SHORT G_M38470_IG06
       ;...
       test     rcx, rcx
       je       SHORT G_M38470_IG07
```
If you simplify it to
```C#
&& (comparer != null ? comparer.Equals(candidateEntry.key, key) : EqualityComparer<TKey>.Default.Equals(candidateEntry.key, key)
```
it  will remove the second check.